### PR TITLE
Snapsafe-type uniqueness breaking event detection

### PR DESCRIPTION
**Pull request by justsmth**
### Issues:
* Resolves: CryptoAlg-2505

### Description of changes: 
* Add support for Snapsafe-type uniqueness breaking event detection on Linux using SysGenID (`/dev/sysgenid`). Adds 4 new (internal-use) functions:
  * `int CRYPTO_get_snapsafe_generation(uint32_t *snapsafe_generation_number);`
  * `int CRYPTO_get_snapsafe_active(void);`
  * `int CRYPTO_get_snapsafe_supported(void);`
  * `const char *CRYPTO_get_sysgenid_path(void);`

### Call-outs:
New build flags added:
* `TEST_SYSGENID_PATH` -- a temporary file path that will be used for testing. The user executing the test must have permission read/write to this path and it must be created prior to running the tests.

### Testing:
* Added new "SnapsafeGenerationTest" test-suite. The test logic differs by platform.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: crypto/fipsmodule/rand/rand.c
Comment (andrewhop) [inline crypto/fipsmodule/rand/rand.c:410]: I realize you didn't add this but you are just continuing the pattern: why the memset 0? Even if we have fork protection or snapsafe why not just mix in whatever uninitialized value was in memory?

File: crypto/fipsmodule/rand/rand.c
Comment (andrewhop) [inline crypto/fipsmodule/rand/rand.c:557]: Should this check the return value of CRYPTO_get_snapsafe_generation?

File: crypto/fipsmodule/rand/snapsafe_detect.c
Comment (andrewhop) [inline crypto/fipsmodule/rand/snapsafe_detect.c:30]: Should this be `R_OK` in the unlikely case it exists but we can't access it?

File: crypto/fipsmodule/rand/snapsafe_detect.c
Comment (andrewhop) [inline crypto/fipsmodule/rand/snapsafe_detect.c:58]: Isn't this always true since this function sets it to null on line 43? Does it make sense to check if sgc_addr_bss_get is null at the start of the function so this is a lazy init that only happens once?

File: crypto/fipsmodule/rand/snapsafe_detect.c
Comment (andrewhop) [inline crypto/fipsmodule/rand/snapsafe_detect.c:92]: Just double checking are we reading the right part of the mmap data? From the documentation:
> The driver supports ``PROT_READ, MAP_SHARED`` mmaps of a single page
> in size. The first 4 bytes of the mapped page will contain an
>  up-to-date u32 copy of the system generation counter.

So the pointer points to the page of memory and the counter is at address -> address + 3, not something like an endian issue and the counter is on the other side of the page.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (andrewhop) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:25]: What is 0600? Cna that use any of the other flag values to make it clearer?

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (andrewhop) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:69]: ASSERTS will fail and stop executing the test and is our normal test convention.
```suggestion
  ASSERT_TRUE(init_sgn_file(&sgn_test)));
```

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (andrewhop) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:96]: NP: this could use an EXPECT_EQ so all values would be executed and you could see if it's always wrong or just wrong for some of them:
```suggestion
    EXPECT_EQ(new_sysgenid_value_hint, current_snapsafe_gen_num);
```

File: crypto/fipsmodule/rand/snapsafe_detect.c
Comment (justsmth) [reply crypto/fipsmodule/rand/snapsafe_detect.c:30]: For this function, the goal is to determine whether the `/dev/sysgenid` exists regardless of whether we are able to fully access it. We take the existence of `/dev/sysgenid` as implying that the system is capable of "snapshots", and thus we should take extra precaution w.r.t. entropy collection.

File: crypto/fipsmodule/rand/rand.c
Comment (justsmth) [reply crypto/fipsmodule/rand/rand.c:557]: Good point.  We only really care about this when snapsafe detection is active.  (Although when it's inactive, the generation value wouldn't change anyways.) I'll make this logic also conditioned on `CRYPTO_get_snapsafe_generation` returning 1.

File: crypto/fipsmodule/rand/rand.c
Comment (justsmth) [reply crypto/fipsmodule/rand/rand.c:410]: The memset to 0 is done to avoid collecting extra data/entropy via a slow syscall/rdrand invocation unless it's required. So this change essentially makes it required when snapsafe is supported but not initialized properly.

File: crypto/fipsmodule/rand/snapsafe_detect.c
Comment (justsmth) [reply crypto/fipsmodule/rand/snapsafe_detect.c:92]: As long as we're reading the first (i.e., smallest-address) 4 bytes of the page and interpreting them as an (endian-appropriate) u32 value, the logic should be correct.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (justsmth) [reply crypto/fipsmodule/rand/snapsafe_detect_test.cc:25]: I can change it to [`S_IRUSR | S_IWUSR`](https://en.cppreference.com/w/cpp/filesystem/perms) instead.

File: crypto/fipsmodule/rand/rand.c
Comment (andrewhop) [reply crypto/fipsmodule/rand/rand.c:410]: Well that's just what this branch does, you could avoid collecting slow system entropy *and* avoid calling memset. Nothing I see in the rest of this function depends on this additional_data being all 0 if there is no system entropy to collect.

File: BUILDING.md
Comment (torben-hansen) [inline BUILDING.md:233]:
```suggestion
## Snapsafe Prerequisites
```

File: CMakeLists.txt
Comment (torben-hansen) [inline CMakeLists.txt:105]: Is this necessary? You can define `AWSLC_SYSGENID_PATH` in your build if you want to.

File: crypto/fipsmodule/bcm.c
Comment (torben-hansen) [inline crypto/fipsmodule/bcm.c:131]: Order alphabetically

File: util/build_compilation_database.sh
Comment (torben-hansen) [inline util/build_compilation_database.sh:15]: looks a bit unrelated?

File: util/all_tests.json
Comment (torben-hansen) [inline util/all_tests.json:59]: should try Arm as well

File: tests/ci/run_posix_tests.sh
Comment (torben-hansen) [inline tests/ci/run_posix_tests.sh:48]: does this pass through memory testers?

File: crypto/fipsmodule/rand/rand.c
Comment (torben-hansen) [inline crypto/fipsmodule/rand/rand.c:105]: could replicate `fork_generation` and simply name it `snapsafe_generation`?

File: crypto/fipsmodule/rand/rand.c
Comment (torben-hansen) [inline crypto/fipsmodule/rand/rand.c:104]:
```suggestion
  // the drbg state must be reseeded.
```

File: crypto/fipsmodule/rand/rand.c
Comment (torben-hansen) [inline crypto/fipsmodule/rand/rand.c:556]: Hmm this could be replaced by:
```suggestion
  if (1 == CRYPTO_get_snapsafe_generation(&snapsafe_generation_id)) {
```
and just always do the test. If snapsafe is not supported in runtime, the test will always pass.

Current impl saves a few cycles, which I guess is your motivation?

File: crypto/fipsmodule/rand/rand.c
Comment (torben-hansen) [inline crypto/fipsmodule/rand/rand.c:559]: maybe needs a bit of elaboration.

**Review by torben-hansen — COMMENTED** [review]
Still need to review `crypto/fipsmodule/rand/snapsafe_*`

File: tests/ci/run_posix_tests.sh
Comment (torben-hansen) [reply tests/ci/run_posix_tests.sh:48]: In fact, this doesn't actually run any tests AFAIK

Comment (justsmth) [reply tests/ci/run_posix_tests.sh:48]: I think all of Linux-based our CodeBuild jobs use it.

File: crypto/fipsmodule/rand/snapsafe_detect.c
Comment (andrewhop) [inline crypto/fipsmodule/rand/snapsafe_detect.c:21]: Does this need to be a variable? Can the two places that use the file path use AWSLC_SYSGENID_PATH directly?

**Review by andrewhop — COMMENTED** [review]
Core change looks good to me, ship it pending urandom test fix.

File: crypto/fipsmodule/rand/snapsafe_detect.c
Comment (justsmth) [reply crypto/fipsmodule/rand/snapsafe_detect.c]: Good point.  There's no need for a variable.

File: CMakeLists.txt
Comment (justsmth) [reply CMakeLists.txt:105]: Ok, I'll remove this build option.

File: util/all_tests.json
Comment (justsmth) [reply util/all_tests.json:59]: Ok, I'll just remove the `target_arch`.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:101]:
```suggestion
  if (CRYPTO_get_snapsafe_supported()) {
```

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:70]: can't use `ASSERT_TRUE`?

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:79]:
```suggestion
  ASSERT_EQ((unsigned int) 7, current_snapsafe_gen_num);
```

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:23]: can be static.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:57]: static?

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:26]: can use the `ASSERT_` and `EXPECT_` in these utility functions as well.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:29]: missing newline before if-statement.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:41]:
```suggestion
  size_t pgsize = (size_t) page_size;
```

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:34]: just a suggestion: could zeroise the entire memory area

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:48]: not an issue for the testing. But the file descriptor leaks if the four previous if-statements are true.

Comment (torben-hansen) [reply crypto/fipsmodule/rand/snapsafe_detect_test.cc:48]: Our static analyser might complain after merging this in though...

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:18]:
```suggestion
  void *addr;
```

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:59]: a'ha I think I missed this step back when I tried making this file-mocking method work....

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:58]:
```suggestion
  memcpy(sgn_test->addr, &val, sizeof(uint32_t));
```

File: crypto/fipsmodule/rand/rand.c
Comment (torben-hansen) [inline crypto/fipsmodule/rand/rand.c:104]:
```suggestion
  // snapsafe_generation is non-zero when active. When the value changes,
```

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (justsmth) [reply crypto/fipsmodule/rand/snapsafe_detect_test.cc:62]: Will fix.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (justsmth) [reply crypto/fipsmodule/rand/snapsafe_detect_test.cc:29]: But if we do that we would be unable to cleanup resources, e.g., close the file. I could use `FAIL()` instead of returning 0?

Comment (torben-hansen) [reply crypto/fipsmodule/rand/snapsafe_detect_test.cc:29]: OK that makes sense.

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (WillChilds-Klein) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:42]: nit: `close(fd_sgn)` before returning?

File: crypto/fipsmodule/rand/rand.c
Comment (WillChilds-Klein) [inline crypto/fipsmodule/rand/rand.c:391]: what if this is 0 after the call to `CRYPTO_get_snapsafe_generation`? is that a case we need to handle as on L558 or retry?

File: crypto/fipsmodule/rand/snapsafe_detect_test.cc
Comment (WillChilds-Klein) [inline crypto/fipsmodule/rand/snapsafe_detect_test.cc:36]: nit: why call `fsync` here? we haven't buffered any data

Comment (justsmth) [reply crypto/fipsmodule/rand/snapsafe_detect_test.cc:36]: Good point. Will remove.

File: crypto/fipsmodule/rand/rand.c
Comment (justsmth) [reply crypto/fipsmodule/rand/rand.c:391]: It's expected that `snapsafe_generation` will likely be 0 after the call to `CRYPTO_get_snapsafe_generation` -- in fact, that'll be true for any environment that doesn't have the `/dev/sysgenid` setup in the kernel.  What we care about is whether the `snapsafe_generation` has *changed* since the last time we re-seeded the DRBG.

File: crypto/fipsmodule/rand/snapsafe_detect.h
Comment (torben-hansen) [inline crypto/fipsmodule/rand/snapsafe_detect.h:51]:
```suggestion
// On success, it returns 1. Otherwise, returns 0.
```

File: crypto/internal.h
Comment (torben-hansen) [inline crypto/internal.h:118]: what's this needed for?

Comment (justsmth) [reply crypto/internal.h:118]: I think it was so that gtest main could call the HAZMAT function.

File: crypto/fipsmodule/rand/rand.c
Comment (torben-hansen) [reply crypto/fipsmodule/rand/rand.c:104]: Could still resolve this typo.

**Review by torben-hansen — APPROVED** [review]

**Review by torben-hansen — APPROVED** [review]

**Review by dkostic — APPROVED** [review]

**Pull request merged**
